### DISCUSSION TITLE
Error Handling for Mutations

### DISCUSSION THREAD

**Discussion by rubber-duck-software** [remix-run/remix#5957] [Proposals]
# Proposal

Allow action errors to be handled without unmounting the route's default component.

## Background

In a traditional SPA, a very common error handling pattern for mutations is something like this

"If the mutation throws an error, we show the user a dialog which informs them of the failure. When the dialog is dismissed, the user should be seeing their form again still filled out and ready to retry the submission."

This is very user friendly (and relatively common across the web). In a more traditional React SPA, the code to make this happen might look something like 

```js
<form onSubmit={async event => {
    event.preventDefault()
    try {
       await fetch('/some-route', { method: 'post' })
    } catch (error) {
       showErrorDialog()
    }
}}>
...
</form>
```

Accomplishing this same behavior is Remix is currently not possible to my knowledge. Let's say that our user is experiencing network issues when they try to submit a Remix Form. The server will never receive the request and an error will be thrown. There is no opportunity to catch this error in user-land and so Remix renders the ErrorBoundary. 

But wait! When we rendered the error boundary, we unmounted the default route component losing all our form data and react state! This is not a good user experience and there are not good workarounds available. 

## Proposal Details
I propose that the ErrorBoundary should render if there in an error on component render or in the loader for the route, but not if there is an error in the action. The reason for the distinction is that even if the action fails, we can go on rendering the default route component. It still has all it needs to render, so why unmount it? 

The UI for an action error should be able to be "added on top of" the existing UI.

## Why does this belong in Remix
The Remix docs talk about using progressive enhancement to ensure your UI is resilient to network issues. See https://remix.run/docs/en/1.14.3/tutorials/blog#progressive-enhancement 

The Remix docs talk about maintaining form state in case of server error (does not mention network errors). See https://remix.run/docs/en/1.14.3/guides/optimistic-ui#maintain-form-state

I think allowing users to maintain form state while being resilient to network issues is a very natural fit. 

## Possible APIs
These are just my initial thoughts
1. useActionError
```js
export default function MyRoute() {
    const actionError = useActionError()
    
    return (
        <div>
            <Form>
                ...
            </Form>
            {actionError ? <span>sorry try again</span> : null}
       </div>
    )
}
```
2. ActionErrorOutlet
```js
export default function MyRoute() {
    return (
        <div>
            <Form>
                ...
            </Form>
            <ActionErrorOutlet />
       </div>
    )
}
```

**Comment by kiliman** [comment]
I commented on your previous post before reading this proposal. I like it. I think it should be opt-in, though. Something like:
```ts
<Form handleError />
```

**Reply by rubber-duck-software** [reply]
linking to previous post for convenience https://github.com/remix-run/remix/discussions/5955

**Reply by kentcdodds** [reply]
I don't think we'd want to do that as it wouldn't progressively enhance. We'll want to make sure have something that would work with/without JS.

**Reply by kiliman** [reply]
Yes, I understand. The issue is that if there's an error, it will always render the `ErrorBoundary` even if it's recoverable. For non-JS scenarios, that's fine as that's how the Web has always worked. However, this **IS** progressive enhancement, in that with JS, we can opt-in to more granular error handling.

**Reply by rubber-duck-software** [reply]
Quick Table to make sure we're all on the same page regarding the proposal

### Current Remix Behavior
| | Server Returns non-200 | Server throws Error | Network Error |
|--------|--------|--------|--------|
| With JS|  Render Default Component | Render Error Boundary | Render Error Boundary |
| Without JS | Render Default Component | Render Error Boundary | Render Default Browser Network Error Page | 

### Proposed Behavior
| | Server Returns non-200 | Server Throws Error | Network Error |
|--------|--------|--------|--------|
| With JS| Render Default Component | Custom Error Handling UI | Custom Error Handling UI |
| Without JS | Render Default Component | ? | Render Default Browser Network Error Page | 

@kentcdodds - If I'm not mistaken, you're pointing out that we need to consider how the custom error handling works if the site is running without JS and we encounter a server error.

In my mind, we can go one of two directions with this. 

1. We could consider users opting-in to custom error handling a progressive enhancement. In this case, if there is a server error and the site is running without JS, we would have to render the nearest error boundary (I think this is what @kiliman is thinking anyway)
2. We can devise an api which has enough flexibility to render the default component alongside an error component for a particular route. This strikes me as much more complicated to get right.

**Reply by sergiodxa** [reply]
> if there's an error, it will always render the `ErrorBoundary` even if it's recoverable.

If there's an Error being throw from the action that's unexpected, and the error boundary should be rendered IMO.

If you know the form submit can fail, e.g. because of an invalid value, you should return a response with a 4xx status code, then your normal component UI will remain mounted and you can handle the return from useActionData.

If there's a change to be made here could be to don't render CatchBoundary if the action throw a non-redirect response, so I could write my action like this:

```ts
export async function action({ request }: DataFunctionArgs) {
  let formData = await request.formData()
  let { status, errors } = await validate(formData)
  if (status === "invalid") throw json({ errors }, 400) // throw a bad request response
  let { status, errors, result } = await performMutation(formData)
  if (status === "failure") throw json({ errors }, 400) // throw a bad request response
  return json(result)
}
```

Right now, those `throw json({ errors }, 400)` go to the CatchBoundary, replacing the form, which is usually not what you want.

But if an error happened, for example when reading the request body as FormData, this is unexpected and it's ok to render the ErrorBoundary.

This will make the ErrorBoundary essentially a UI only for 500 responses where the server had an unexpected error that it didn't know how to handle (the definition of a 500 response).

And if you wanted to catch any possible error you could wrap your code in a try/catch.

```ts
export async function action({ request }: DataFunctionArgs) {
  try {
    // same code as above
  } catch (exception: unknown) {
    if (exception instanceof Response) throw exception // re-throw responses

    // probably handle specific errors here, like TypeError, ZodError, etc.

    // report unexpected exceptions to Sentry or something similar so you don't hide the error
    Sentry.captureException(exception)

    // if it's an Error, throw a 400 response with the error.message
    if (exception instanceof Error) throw json({ errors: [exception.message] }, 400)

    // for any other thrown value (e.g. strings, numbers, objects, etc.) let ErrorBoundary render
    throw exception
  }
}
```

**Reply by kentcdodds** [reply]
I think the primary use case to consider here is if there's an error fetching due to issues around network reliability. It's always bothered me that I don't know what to do if the user submits a form in a spotty network situation and they just happen to submit it when the network drops. Pretty sure that'll go to the error boundary. I think in that situation it would be better to have an alternative.

In that scenario, this is definitely progressive enhancement as @kiliman suggested because without JS you'll get a full-page refresh and the browser will fail to communicate with the server and the whole page will disappear anyway. So maybe sticking something on the form (and an option in the fetcher) would be appropriate.

**Reply by lili21** [reply]
@sergiodxa what's about network issue? the error is not throw by action.

**Reply by sergiodxa** [reply]
I think network errors are an special case, first of all, a network error happen only client-side, not in your action (you could have network errors but that would be between your action and a 3P and you can handle them yourself).

Maybe Remix could provide a way to know about the network status, starting with being online or offline, and a way to know when a network error happened in the app.

```ts
// it can help you disable form submits or don't do it if the user is offline
let isOnline = useNetworkStatus()

// use isOnline to prevent the submit
function handleSubmit(event: FormEvent<HTMLFormElement>) {
  if (isOnline) return
  event.preventDefault()
  showErrorDialog()
}

// if a network error happened, you can know it here, the next form submit must reset this
let networkError = useNetworkError()

// use networkError to open the error dialog
useEffect(() => {
  if (networkError) showErrorDialog()
}, [networkError, showErrorDialog])

return (
  <Form method="post" onSubmit={handleSubmit}>
    {!isOnline && <NetworkOfflineUI />} // tell the user that it's offline
    {/* render inputs here */}
    <button disabled={!isOnline}>Submit</button> // disable submit if offline
  </Form>
)
```

The network status is something that you could build yourself easily but the useNetworkError is something you definitely need deep integration with Remix so they can catch those error and give them back to you instead of showing the error boundary.

The useNetworkStatus is useful tho, you can prevent the submit event to happen and render something, you can disable the submit button, you could even show a UI to tell the user they're offline, you could event prevent the submit and save the data in localStorage, ask the user.

Ideally you want to handle networks error before the user submits the form, so you don't even do the network request, but if something happens after you could leverage useNetworkError to handle the error when it happens.

It could be interesting to somehow know if there's at least one instance of `useNetworkError` then let the developer handle them, if not throw the error so the ErrorBoundary renders.

**Reply by lili21** [reply]
@sergiodxa I guess there will be many edge-cases. Network is hard to predicate. you can be online when you click the button and off after that, before you get the response.

**Reply by sergiodxa** [reply]
That’s where useNetworkError comes to play, it can give you the error when that happens.

**Reply by kentcdodds** [reply]
Those hooks seem like a pretty nice solution to this. Would definitely need a change in the router for this. Anyone want to put together an RFC?

**Reply by rubber-duck-software** [reply]
I love the idea of `useNetworkStatus` but, I want to advocate for a solution different than `useNetworkError` and closer to `useActionError`. 

Error handling in Remix has at least 3 dimensions.

1. Where did the error happen? (component, loader, or action)
2. What type of error is it? (Server returns error response, server throws error, or network error)
3. Do we have javascript available? 

The question of whether we use something like `useNetworkError` or `useActionError` comes down to how we want users to think about network errors.

I think there are two options available

1. Network errors when calling an action function should be handled in the same way as network errors when calling a loader function.

This approach compresses the **Where** dimension of error handling. Whether the network error happened when calling loaders or when calling actions, we use the same API the handle the error. This is the approach suggested by `useNetworkError`.

2. Network errors when accessing the server should be handled in the same way as errors thrown by the server. 

This approach compresses the **What type** dimension of error handling. Whether the error is a network error or an error thrown by the server, we would use the same API to handle the error. This is the approach suggested by `useActionError`


@sergiodxa, I know you are far closer to the problem than myself, but all the same, I want to give you a few things to think about.

**Reply by sergiodxa** [reply]
I think `useActionError` will encourage throwing custom Error instances from your action, but an Error should be limited to unexpected scenarios, even if you throw the error in your code, the idea of `throw error` should (IMO) be "this shouldn't have happened, and I can't recover from it" so there rendering ErrorBoundary is ok, you don't know what to do and fallback to an error UI.

Meanwhile it's a known possibility the moment we have a network between the client and the server there can be network errors, and a network error could also happen when fetching a loader not just when submitting to an action, so `useNetworkError` could also work for loaders, you could have a single `useNetworkError` at the root that triggers a toast notification "Something failed, check your connection", or you could have many and show on the form that the request failed due a network error.

There's also another possibility, your request reach the server and was processed, but the response failed to be received, or the revalidation of loaders failed, in those cases for example do you want the user to retry the form? Maybe trigger a new revalidation with useRevalidator?

Allowing any error to be thrown from the `action` would make `useActionError` too generic in my mind, you would have to fill your components using the hook with `if (actionError instanceof SomeError)`, and I imagine serializing custom error classes like ZodError will be way harder and way more code.

**Reply by rubber-duck-software** [reply]
> I think useActionError will encourage throwing custom Error instances from your action

😅 Guilty as charged. This is exactly what I was planning. ZodError serializes surprising well (a ZodError can be constructed from the list of issues which were serialized over the wire).

Honestly, I am irritated with having to wrap all my action functions with a try catch block and a custom error handling function. And since higher order functions aren't allowed (https://remix.run/docs/en/1.15.0/guides/constraints#higher-order-functions) this logic cannot be abstracted for easy reuse. 

You're right. useActionError attempts to address a second pain point I am experiencing around error handling in Remix, but that does not mean it is the right API for the network error use case. 

I would say that the useNetwork error should somehow expose where the network error happened. Users might handle the error differently if it happened trying to call the loaders or if it happened trying to call an action. Additionally there should be a way to "fallback" to default error handling. A user should be able to handle all network errors when calling actions, but use the default error handling if a network error happens when calling a loader.

**Reply by sergiodxa** [reply]
> I am irritated with having to wrap all my action functions with a try catch block and a custom error handling function

In Zod case, you could use `schema.safeParse` to don't get thrown errors

```ts
let result = schema.safeParse()
if (!result.success) {
  // use result.error here, it's a ZodError
}
// use result.data here, it's the validated schema
```

> useNetworkError should somehow expose where the network error happened

I agree, I'm not sure what should be the exact return type of the hook, it would have to be discussed by the team, I think know if the error happened on a request to a loader or action is important, probably also the URL that was requested, and the formData sent if it's an action.

> Additionally there should be a way to "fallback" to default error handling

You could throw an error yourself:

```ts
let networkError = useNetworkError()
if (networkError) throw new Error("Network failed") // this will trigger the ErrorBoundary
```

I'm not sure if it's possible to know if no instances of the hooks are active, but if it's possible Remix could fallback to always throw if there's no `useNetworkError()` call.

**Reply by lili21** [reply]
@sergiodxa Let's say I'm calling another backend service in my action, and we all know that the call won't be 100% successful. If an error occurs that we can't handle, instead of showing an error UI, we typically want to provide a hint to the user and let them try again.

**Reply by lili21** [reply]
and actually in most case, at least in our products, that's what we need - if user trigger an action and the action failed, no matter why it's failed, we just want keep the page/ui, and show the error message nicely.

**Reply by kentcdodds** [reply]
Right, and in those cases you get to manage that in your action so you can send a response and get it in the useActionData.

Again, I'm pretty sure this issue is only really about handling when making a network call to the action fails. Situations in which our action code is never even called and therefore our action code can't provide error messages.

**Reply by lili21** [reply]
to do this, we have to using try catch in every action ,

```js
export const action = async () => {
  try {
    await callAnotherService();
    return json({ success: true });
  } catch (e) {
    return json({ success: false, reason: e.message }, { status: 500 });
  }
}
```

and don't forget the status code option, cause in production, we need to know how many api call are failed and do some metrics and alerting.

in this way, why don't just let the action throw?

**Reply by lili21** [reply]
I think it's the same problem, what should we do to handle the action exception?
is there any case the we want the page/ui fallback to ErrorBoundary/CatchBoundary because of user click a submit button?

**Reply by rubber-duck-software** [reply]
The original nature of this proposal was 
> Allow action errors to be handled without unmounting the route's default component.

Which is currently impossible in case of a network error and (I believe) not very ergonomic in case of an error thrown in the action. At first, I was expecting the same API change to handle both of these cases, but am now realizing that these cases could be handled separately. 

Assuming that we go forward with `useNetworkError` for handling network errors, I have created a new discussion for improving the ergonomics of handling errors thrown in actions. See #6008. 

To ensure the API for handling network errors ends up in the right place, I think we should (at least briefly) indulge a discussion about making handling action errors more ergonomic. It is possible, but not necessary, that these apis are related.

**Reply by lili21** [reply]
maybe we can do it with `fetcher.submit` or `submit` return from `useSubmit`?. 

Using the `<Form />` if you want something more like browser behavior and works even it's not hydrated yet, and when the action failed, the exception captured by boundary.

Using `fetcher.submit` if you want do it in another way,  we can provide a promise based API 

```js
fetcher.submit()
  .then(() => {
  })
  .catch(() => {
    // handle the exception, it can be the network error, or the action just failed.
  });
```

**Comment by rubber-duck-software** [comment]
@kiliman - let me see if I understand you. Would the api be something like below?
```js
<Form handleError={e => /* handle the error */} />
```

and then the default implementation for `handleError` would be to simply throw the error. 

And for the use case I described above would be implemented as
```js
<Form handleError={e => showErrorDialog()} />
```

**Reply by kiliman** [reply]
Yeah, I'm not keen on the name. I probably shouldn't have used "handle" since I'm not saying it should be an event handler, but simply a marker that I opt-in to manually handling errors via `useActionError` or `ActionErrorOutlet`. In other words, the default should be how Remix currently handles errors.

**Reply by rubber-duck-software** [reply]
As I think about it, why not an event handler? An event handler might be the most flexible for developers to use. If developers were given an event handler, they could easily implement their own versions of `useActionError` or `ActionErrorOutlet`. 

But, honestly, I'm not familiar enough with the Remix code base to know what is possible. I could also see 
```js
<Form errorHandling="throw" /> // the default
<Form errorHandling="manual" /> // opting into custom error handling UI
```

**Comment by lili21** [comment]
we ran into the same issue so many times.

**Reply by lili21** [reply]
we need this.

**Comment by lili21** [comment]
especial for `fetcher.load` and `fetcher.submit`, no reason this fetching error should unmount entire route.

**Comment by scolestock** [comment]
Planned as a possibility anytime soon?

**Comment by machour** [comment]
Linking to https://github.com/remix-run/react-router/discussions/10013 as it also addresses the situation

**Comment by chrisvltn** [comment]
This would be amazing to have. A failure sending a response in a pop-up chat in the corner of the page shouldn't break the whole page. :)

**Comment by okalil** [comment]
I stopped using actions because of that. But now that clientAction landed, we can probably handle network errors there.

**Comment by okalil** [comment]
I imagine something like that should work
```tsx
export const clientAction = async ({ serverAction }) => {
  try {
    return await serverAction()
  } catch (error) {
    return handleNetworkError(error)
  }
}
```

**Reply by okalil** [reply]
`handleNetworkError` could trigger the dialog/toast for instance since the function runs on client side

**Comment by gitnik** [comment]
Is there any news regarding this? It's a major pain point for us.

**Comment by brophdawg11** [comment]
This discussion is being closed as part of the Remix v3 repository cleanup.

Remix v1/v2 was [upstreamed into React Router v7](https://remix.run/blog/wake-up-remix), so if this proposal is still relevant, please reopen or continue the discussion in the [React Router](https://github.com/remix-run/react-router) repository.
